analyzers/src/SonarAnalyzer.Common/Rules/Utilities/SymbolReferenceAnalyzerBase.cs

```diff
@@ -139,9 +142,13 @@ private static SymbolReferenceInfo.Types.SymbolReference GetSymbolReference(IRea
         {
             for (var i = 0; i < references.Count; i++)
```

Comment:
I have changed to `foreach`.
However, it's worth mentioning that the `for` has been introduced in a task to improve the performance of the `SymbolReferenceAnalyzer`: https://github.com/SonarSource/sonar-dotnet/pull/5548
I hope the `foreach` is optimized enough to not impact on the overall performance of this utility analyzer, and that this loop is not the hot path.

Reply:
I think the implementation uses `IList<T>` which is a bit unfortunate. it would be better to work with `ImmutableList<T>` or `ImmutableArray<T>` instead as for these, the compiler picks `struct` enumerable which are allocation free. For `IList<T>` this is not possible and a class bases enumerable is returned.
[sharplab.io](https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLQGED2AbbEDGMAlpgHYAyRMECU2yAPgAIBMAjALABQTADAARM2AFgDc3PoLYA6LLgLEyyaQEkAtmrgwowPOJ4BmQS37p+Ab279rgo02H8AsgAoVlZDAA8RUjAB8/NhEHgCUFlY2kWCYSFD4ABbOAG5QCPxE6aSBwTAhEZHWllwFJVIAnM5EIfql/AC++fWNjUx2Di7uXj7+2aHhxaXRsQnJqemZvbmNkUW1kUIVVTWlDQPWdZHNa7aC7a4aWjp4nd6+AUF9s4Mx0CMpaRk+k3nbM9OlC5XV7+uNGzbcOpAA==)

Reply:
See e.g. ImmutableArray has two implementations for the Enumerator:
https://github.com/tunnelvisionlabs/dotnet-collections/blob/master/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray%601%2BEnumerator.cs

One is used if you foreach over the ImmutableArray directly and the other one, if you foreach through an interface over it.

Reply:
As discussed, I have extracted this change into a new, dedicated issue: https://github.com/SonarSource/sonar-dotnet/issues/8627, since in my opinion the change is more structural, performance-sensitive and too important to be in the scope of this issue.